cmd/srcd/config/config.go

```diff
@@ -1,83 +1,71 @@
 package config
 
 import (
+	"io/ioutil"
 	"path/filepath"
 
+	"github.com/src-d/engine/api"
+
 	homedir "github.com/mitchellh/go-homedir"
 	"github.com/pkg/errors"
-	"github.com/sirupsen/logrus"
-	"github.com/spf13/viper"
-	"github.com/src-d/engine/api"
+	"gopkg.in/src-d/go-log.v1"
 	yaml "gopkg.in/yaml.v2"
 )
 
-// InitConfig reads the config file and ENV variables if set.
-func InitConfig(cfgFile string) {
+var configFile string
+
+// InitConfig reads the config file if set.
+func InitConfig(cfgFile string) error {
 	if cfgFile != "" {
 		// Use config file from the flag.
-		viper.SetConfigFile(cfgFile)
+		configFile = cfgFile
 	} else {
 		// Find home directory.
 		home, err := homedir.Dir()
 		if err != nil {
-			logrus.Fatalf("Could not detect home directory: %s", err)
+			return errors.Wrapf(err, "Could not detect home directory")
 		}
 
-		// Search config in $HOME/.srcd/ with name "config" (without extension).
-		viper.AddConfigPath(filepath.Join(home, ".srcd"))
-		viper.SetConfigName("config")
-	}
-
-	viper.AutomaticEnv() // read in environment variables that match
-
-	// If a config file is found, read it in.
-	err := viper.ReadInConfig()
-	if err != nil {
-		// viper returns ConfigFileNotFoundError only when the default config file
-		// is not found. For a wrong file path on --config it's an os.PathError
-		if _, ok := err.(viper.ConfigFileNotFoundError); !ok {
-			logrus.Fatalf("Error reading the config file %s: %s", viper.ConfigFileUsed(), err.Error())
-		}
+		configFile = filepath.Join(home, ".srcd", "config.yml")
 	}
 
-	checkConfig()
+	return checkConfig()
 }
 
-func checkConfig() {
-	actualConfigFile := viper.ConfigFileUsed()
-	if actualConfigFile == "" {
-		return
+func checkConfig() error {
+	if configFile == "" {
+		return nil
 	}
 
-	logrus.Debugf("Using config file: %s", actualConfigFile)
+	log.Debugf("Using config file: %s", configFile)
 
 	// The config file may define an int field as string, or have extra fields.
 	// Calling Config we force a check on initialization.
 	_, err := Config()
 	if err != nil {
-		logrus.Fatal(errors.Wrapf(err, "Error checking config file '%s'",
-			actualConfigFile))
+		return errors.Wrapf(err, "Error checking config file '%s'", configFile)
 	}
+
+	return nil
 }
 
-// Config returns the config used (from a file, env, or defaults)
+// Config returns the config used
 func Config() (*api.Config, error) {
```

Comment:
This is called only [here](https://github.com/src-d/engine/blob/87b3d351b370d6d41065a64536d2afb9d4c6a6c7/cmd/srcd/daemon/daemon.go#L141), but it maybe called multiple times given that we may call `ensureStarted` more than once. So we actually unmarshall the configuration every time we need it.

What about having a global `var conf api.Config`? I'd unmarshall the config only once at `InitConfig` and then add functioin to get that config.

Reply:
I have refactored the file in 51122b2, I think now it's closer to what you describe.